### DISCUSSION TITLE
Efficiently updating robot knowledge in dynamic real-world scenarios

### DISCUSSION THREAD

**Discussion by MannavaVivek** [aiplan4eu/unified-planning#469] [Q&A]
To define my usecase, I want to use this planning framework in an real world mobile robot and once it generates the plan, start executing the actions and adding that knowledge/effects to the knowledge base (which I believe would be the problem itself in this case) and when the robot encounters an issue, invoke the replanner with the latest information on the various objects around the robot to continue the operation. 

Can you suggest a better alternative to setting each value to false and setting a new value to true? Since I am getting the original problem from a pddl file, the initial values are all in the format of at(robot, start) := true, I am currently doing at(robot, start) := false, at(robot, loc1) := true  and if I have to move the robot between multiple locations, this quickly adds up to a lot of initial value conditions.

**Comment by mikand** [comment]
Hi @MannavaVivek!

Not sure if I completely got your problem, but you can either (1) use python loops to set the values or (2) use a variable-value representation (but this second case is not supported in PDDL).

1. You can use python code to set the initial state as needed
```python
real_location_name = "l1"
for l in all_location_objects:
    problem.set_initial_value(robot_at(l), l.name == real_location_name)
```
In this example code, you set the whole initial state (the truth values of all `robot_at`) using python

2. You can model your problem using a variable-value representation. Instead of using a `robot_at` predicate, you can create a fluent of type `Location` called `position`:
```python
position = Fluent('position', Location)
[...]
problem.set_initial_value(position, real_location)
``` 
The drawback of this approach is that you cannot model your problem in PDDL and few solvers support this. However, you can compile this feature away with a `Compiler`. 

Is this what you were after?

**Reply by MannavaVivek** [reply]
Not quite. I tried to explain it better here:

I have an action like so which was generated from the pddl file
```
  action move_base(robot r, location source, location destination) {
    preconditions = [
      (at(r, source) and gripper_is_free(r))
    ]
    effects = [
      at(r, source) := false
      at(r, destination) := true
      perceived(source) := false
    ]
  }
```

I want to add the effects to the initial values in real time, so if the robot gets stuck and I have to replan, I only generate a plan for the remainder of the task and not start from the beginning. To do so, for a move_base(location_1,location2) I am doing something like 
replanner.update_initial_value(at(robot, location_1), False)
replanner.update_initial_value(at(robot, location_2), True)

and then for another movebase(location_2, location_3) I would need to change the initial values again, 
replanner.update_initial_value(at(robot, location_2), False)
replanner.update_initial_value(at(robot, location_3), True)

while for each action, there are only two lines, it is my understanding that the replanner would still check the value of at(robot, location_1) during replanning and if I have ten such initial values set to false, i would lose the computation time it takes to check all of those and ensure they are false. 


So I wanted to know if there was any way to do this efficiently. implementing a separate compiler that will take care of the current location might work, but at(robot, location) is not the only fluent I have to be updated as so, and it quickly becomes too many computations.

**Reply by mikand** [reply]
Ok, a bit clearer, but I still do not get why you say that the number of calls to `replanner.update_initial_value` depends on the numer of actions: it actually only depends on the number of locations, right?

In general, unless you have a humongous number of locations it should not be computationally problematic to reset all the initla values:
```python
robot_pos = [location where the robot is according to localization]
for l in locations:
    replanner.update_initial_value(at(robot, l), l==robot_pos)
```
setting a location in this way is just changing a value in a datastructure, it should be quite fast (especially compared to replanning time).

Am I missing something?

**Reply by MannavaVivek** [reply]
In my usecase, it is actually possible for the robot to come back to the location it has already visited,so I would change it for every action. And there also other fluents that I need to update for every action, like the objects are that are being manipulated,  a perceived location flag and so on. The number is not really big, but since it is in a competitive environment, looking to save whatever bit of time I can. But if this works, then it should be fine.
